util/singleflight/singleflight.go

```diff
@@ -143,6 +150,86 @@ func (g *Group[K, V]) DoChan(key K, fn func() (V, error)) <-chan Result[V] {
 	return ch
 }
 
+// DoChanContext is like DoChan, but supports context cancelation. The context
+// passed to the fn function is a context that is canceled only when there are
+// no callers waiting on a result (i.e. all callers have canceled their
+// contexts).
+func (g *Group[K, V]) DoChanContext(ctx context.Context, key K, fn func(context.Context) (V, error)) <-chan Result[V] {
+	ch := make(chan Result[V], 1)
+	g.mu.Lock()
+	if g.m == nil {
+		g.m = make(map[K]*call[V])
+	}
+	if c, ok := g.m[key]; ok {
+		c.dups++
+		c.ctxWaiters.Add(1)
+		c.chans = append(c.chans, ch)
+		g.mu.Unlock()
+
+		// Instead of returning the channel directly, we need to track
+		// when the call finishes so we can handle context cancelation.
+		// Do so by creating an final channel that gets the
+		// result and hooking that up to the wait function.
+		final := make(chan Result[V], 1)
+		go g.waitCtx(ctx, c, ch, final)
+		return final
+	}
+	// Create a context that is not canceled when the parent context is,
+	// but otherwise propagates all values.
+	callCtx, callCancel := context.WithCancel(context.WithoutCancel(ctx))
+	c := &call[V]{
+		chans:  []chan<- Result[V]{ch},
+		cancel: callCancel,
+	}
+	c.wg.Add(1)
+	c.ctxWaiters.Add(1) // one caller waiting
+	g.m[key] = c
+	g.mu.Unlock()
+
+	// Wrap our function to provide the context.
+	go g.doCall(c, key, func() (V, error) {
+		return fn(callCtx)
+	})
+
+	final := make(chan Result[V], 1)
+	go g.waitCtx(ctx, c, ch, final)
+	return final
+}
+
+// waitCtx will wait on the provided call to finish, or the context to be done.
+// If the context is done, and this is the last waiter, then the context
+// provided to the underlying function will be canceled.
+func (g *Group[K, V]) waitCtx(ctx context.Context, c *call[V], result <-chan Result[V], output chan<- Result[V]) {
+	var (
+		res Result[V]
+		err error
+	)
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+	case res = <-result:
+	}
```

Comment:
If there's race between the context being canceled and the result being produced, should one take precedence? Alternatively, if the context was already canceled and a result happened to be ready now, what should take precedence?

If we want cancelation to take precedence, we could do:
```go
select {
case <-ctx.Done():
case res = <-result:
}

...

if err := ctx.Err(); err != nil {
	res = Result[V]{Err: err}
}
output <- res
```
Thus, context errors take precedence over results.

Reply:
I think preferring context cancelation makes sense; if the caller gets a value, it's likely that it'll just end up continuing until a later point and then checking cancelation and aborting, which is essentially wasted work.